{"cells":[{"metadata":{"trusted":false},"cell_type":"markdown","source":["<h1><span style=\"color:red\">Please read this very carefully! </span></h1>\n","\n","In order to setup your own experiments, you need to download remote files to your linux disk image in the collaboratory environment. As data for your user account is NOT reset when you close or reload the HBP, you have to be very careful how you organize & structure your data. In order to help you with that we create a unique working directory for each molecular use case you run.\n","\n","Please be also aware that we switch current working directories in this use case. That means that you have to restart and clear all output in order to go back to your starting directory. "]},{"metadata":{"trusted":true,"collapsed":true},"cell_type":"markdown","source":["## Compare the electrostatic potentials surrounding a set of protein isoforms with multipipsa\n","\n","\n","**Aim:** This use case shows how to use the multipipsa tool to calculate the electrostatic potentials surrounding a set of protein isoforms in aqueous solution, then cluster the isoforms by their electrostatic similarity.\n","\n","**Version:** 1.1 (January 2020)\n","\n","**Contributors:**  Neil Bruce, Lukas Adam, Stefan Richter, Rebecca Wade (HITS, Heidelberg, Germany)\n","\n","**Contact:** [mcmsoft@h-its.org](mailto:mcmsoft@h-its.org)\n","\n","**Note:** This notebook has graphical output using nglview. If you use the \"RunAll\" function of the notebook, this graphical output might not appear on your screen. The cell defined to show the output must be visible in the browser during execution."]},{"metadata":{"trusted":true},"cell_type":"markdown","source":["## Setting up your environment"]},{"metadata":{"trusted":true,"collapsed":true},"cell_type":"markdown","source":["### Check that all required python packages are installed and working"]},{"metadata":{"trusted":true},"cell_type":"code","source":["! pip install --upgrade pip\n","! pip uninstall --yes numpy\n","! pip uninstall --yes pandas\n","! pip install pandas>=1.0.1\n","! pip install numpy>=1.16"],"execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":["# Check that required packages are installed\n","! pip install --upgrade \"hbp-service-client\" \n","! pip install wget python-magic\n","! pip install rpy2==2.9.1\n","! pip install setuptools\n","! pip install --extra-index-url https://projects.h-its.org/pypi multipipsa==4.0.10\n","! pip install nglview\n","! mkdir -p ~/.R/lib\n","! grep -qxF 'R_LIBS_USER=~/.R/lib/' ~/.Renviron || echo 'R_LIBS_USER=~/.R/lib' >> ~/.Renviron\n","! wget -c https://cran.r-project.org/src/contrib/fastcluster_1.1.25.tar.gz\n","! wget -c https://cran.r-project.org/src/contrib/heatmap3_1.1.7.tar.gz\n","! R CMD INSTALL -l ~/.R/lib fastcluster_1.1.25.tar.gz\n","! R CMD INSTALL -l ~/.R/lib heatmap3_1.1.7.tar.gz\n",""],"execution_count":null,"outputs":[]},{"metadata":{"trusted":true,"collapsed":true},"cell_type":"code","source":["# Import python packages/classes used in this notebook\n","import warnings\n","warnings.simplefilter(action='ignore', category=FutureWarning)\n","import numpy\n","import rpy2\n","import os, wget, datetime, magic, inspect\n","from multipipsa.multipipsa import PipsaRun, ApbsRun\n","from multipipsa.clusterpipsa import ClusterPipsa\n","from multipipsa.pipsatypes import DistanceType\n","from PIL import Image\n","from hbp_service_client.storage_service.client import Client\n","import nglview"],"execution_count":null,"outputs":[]},{"metadata":{"trusted":true,"collapsed":true},"cell_type":"markdown","source":["### Set up local directory structure"]},{"metadata":{"trusted":true},"cell_type":"code","source":["# Create a local working directory\n","try:\n","    homeDir = os.environ['HOME']\n","except:\n","    print(\"Error in environment\")\n","\n","else:\n","    workDir = os.path.join(homeDir, 'work')\n","    if not os.path.isdir(workDir):\n","        try:\n","            os.mkdir(workDir)\n","        except:\n","            print(\"unable to make working directory\")\n","    \n","    # Make a new directory to run the use case in. \n","    # If directory already exists, add a number to make a unique name\n","    baseDir = 'wholePIPSA'\n","    dirIter = 0\n","    useCaseDir = os.path.join(workDir, baseDir)\n","    print(useCaseDir)\n","    \n","    if os.path.exists(useCaseDir):\n","        while os.path.exists(useCaseDir):\n","            dirIter += 1\n","            useCaseDir = os.path.join(workDir, baseDir + '.' + str(dirIter))            \n","    \n","    try:\n","        os.mkdir(useCaseDir)\n","    except:\n","        print(\"Failed to make use case working directory\")\n","    else:\n","        print(\"Working directory for current use case: %s\" % useCaseDir)\n",""],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["### Set up collab storage for saving data at end of calculation"]},{"metadata":{"trusted":true},"cell_type":"code","source":["#Find your own collab storage path\n","collab_path = get_collab_storage_path()\n","print(collab_path)\n","storage_client = Client.new(oauth.get_token())"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["# Compare the electrostatic potentials surrounding a set of protein isoforms with multipipsa\n","\n","This use case describes the use of the [multipipsa](https://collab.humanbrainproject.eu/#/collab/19/nav/2108?state=software,multipipsa) software tool to compare the electrostatic potentials of a set of similar proteins. In addition to its own functions, multipipsa makes use of the following open source tools:\n","\n","* [PDB2PQR](https://apbs-pdb2pqr.readthedocs.io/en/latest/pdb2pqr/index.html): A tool that takes a protein structure in [PDB format](http://www.wwpdb.org/documentation/file-format), adds missing hydrogen atoms, and creates a structure file in PQR format. The PQR file format is derived from the PDB format for describing atomic data, but with the occupancy and temperature factor fields replaced with atomic partial charges and radii.\n","\n","* [APBS](https://apbs-pdb2pqr.readthedocs.io/en/latest/apbs/index.html): A tool that calculates electrostatic potentials through solution of the Poisson-Boltzmann equation, one of the most common continuum models for describing electrostatic interactions between molecular solutes in salty, aqueous media. \n","\n","\n","This use case also makes use of the [NGL View](https://github.com/arose/nglview) python package for displaying molecular data. NGL View provides IPython widgets for displaying molecular data inside notebooks, using the [NGL Viewer](http://nglviewer.org/) WebGL molecular viewer.\n","\n","\n","The steps taken in this use case are:\n","\n","* Calculate the electrostatic potentials surrounding a set of similar proteins.\n","* Use PIPSA analysis to calculate the pairwise similarities of the proteins' electrostatic potentials.\n","* Cluster the proteins according to this similarity.\n","\n","\n","For more detailed information on the first step, see the other molecular use case [Calculate the electrostatic potential of a protein from its atomic structure](https://collab.humanbrainproject.eu/#/collab/1655/nav/362934). Instead of comparing the whole electrostatic potentials of the proteins, the potential in a specific region can also be compared. For an example of this see the molecular use case [Compare a specific region of the electrostatic potentials surrounding a set of protein isoforms with multipipsa](https://collab.humanbrainproject.eu/#/collab/1655/nav/362934).\n","\n","## PIPSA analysis\n","\n","PIPSA provides a method for quantitatively comparing the three dimensional interaction property fields of a set of structurally similar proteins. The structures of the proteins must be suitably aligned for the comparison to be reasonable. The method was first described in [Blomberg et al (1999)](https://doi.org/10.1002/&#40;SICI&#41;1097-0134&#40;19991115&#41;37:3%3C379::AID-PROT6%3E3.0.CO;2-K). In this use case, the interaction  property we consider is the electrostatic potential.\n","\n","In the PIPSA method, interaction potential fields are discretised on three dimensional grids, and compared at grid points lying within a skin surrounding the proteins. The skin of the protein is defined as the region begining at a **distance $\\sigma$** from the van-der-Waals surface of the protein's atoms with a **thickness $\\delta$**. Commonly used parameters for this skin are $\\sigma = 3$ Angstrom and $\\delta = 4$ Angstrom. The region to be compared can be further restricted as the intersection of this skin with a sphere or as cone (as shown in the figure below). \n","\n","![PIPSA](https://projects.h-its.org/mcmsoft/pipsa/3.2/f4skin.gif)\n","\n","\n","### Similarity measure\n","\n","The interaction fields of two structures can be compared quantitatively by calculating pairwise similarity indices. One such measure is the Hodgkin index, which is given by \n","$$ SI_{12} = \\frac{2(p_{1},p_{2})}{(p_{1},p_{1})+(p_{2},p_{2})} $$\n","\n","where $ (p_{1},p_{2}) $ is the scalar product of the fields surrounding proteins $1$ and $2$ in the analysis region. They can be calculated as: \n","\n","$$ (p_{1},p_{2}) = \\sum_{i,j,k}\\phi_{1} \\left ( i,j,k \\right )\\phi_{2} \\left ( i,j,k \\right ) $$\n","\n","where $i$, $j$ and $k$ are the three-dimensional spatial coordinates, and $\\phi_{1} \\left ( i,j,k \\right )$ is the potential at point $\\left ( i,j,k \\right )$ on the grid for the template protein. The similarity index runs from $-1$, for completely anti-correlated potentials, to $1$ for identical potentials. At $0$, there is no correlation between the two potentials. The pairwise similarity can be converted to a distance measure, for use in clustering, using\n","\n","$$ D_{12} = \\sqrt{2 - 2SI_{12}} $$\n","\n","\n","\n","\n",""]},{"metadata":{"trusted":false,"collapsed":true},"cell_type":"markdown","source":["## Downloading the protein structures\n","\n","In this use case, we use as our input structures the catalytic domain of nine isoforms of the enzyme adenylyl cyclase (AC1 to AC9), modelled during the work described in [Tong et al (2016)](https://doi.org/10.1002/prot.25167). The following cell downloads these structures from the CSCS storage area."]},{"metadata":{"trusted":true},"cell_type":"code","source":["# Download isoform structure files from CSCS storage for calculation\n","\n","# Loop to download AC1 - 9 structures\n","for iso in range(1, 10):\n","\n","    try:\n","        print(\"Downloading AC%d structure file from CSCS storage area\" % iso)\n","        try:\n","            fileUrl= 'https://object.cscs.ch/v1/AUTH_c0a333ecf7c045809321ce9d9ecdfdea/SGA2_molecular_models/data/Modelled_adenylyl_cyclase_AC_isoform_structures/refined/AC' + str(iso) + '.pdb'\n","        except:\n","            print(\"Error defining file url\")\n","        else:\n","             wget.download(fileUrl, useCaseDir)\n","    except:\n","        print(\"Error downloading structure file AC%d CSCS storage\" % iso)\n","        print(fileUrl)\n","    else:\n","        print(\"Sucessfully downloaded the structure file AC%d from CSCS storage\" % iso)"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["### Viewing the protein structures\n","The following cell creates a molecular viewer to visualise the structures of the AC isoforms. The catalytic domain of AC is a dimer consisting of two protein chains. In the full structure of AC these two chains are connected by a series of transmembrane helices that anchor the protein in the post-synaptic membrane."]},{"metadata":{"trusted":true},"cell_type":"code","source":["# View the downloaded structure\n","# Create a NGL widget object\n","viewPDB = nglview.NGLWidget()\n","# Set the display size\n","viewPDB._remote_call('setSize', target='Widget', args=['600px','400px'])\n","\n","# List of colours to colour each residue differently\n","colorsA=[0xfff5eb,0xfee6ce,0xfdd0a2,0xfdae6b,0xfd8d3c,0xf16913,0xd94801,0xa63603,0x7f2704]\n","colorsB=[0xf7fcf5,0xe5f5e0,0xc7e9c0,0xa1d99b,0x74c476,0x41ab5d,0x238b45,0x006d2c,0x00441b]\n","\n","         \n","# Create list for storing structure components for each isoform. \n","# First item set to None so list index matches isoform number \n","viewPDB_struct = [None]\n","# Loop over isoforms\n","for iso in range(1, 10):\n","    # Define files to load\n","    fname = 'AC' + str(iso) + '.pdb'\n","    AC_struct_file = nglview.FileStructure(os.path.join(useCaseDir, fname))\n","\n","    # Create a component object for displaying the structure of the current isoform\n","    viewPDB_struct.append(viewPDB.add_component(AC_struct_file))\n","    #Clear default representation from the component and add cartoon representations for both chains\n","    viewPDB_struct[iso].clear_representations()\n","    viewPDB_struct[iso].add_representation('cartoon', sele=':A', color=colorsA[9-iso])\n","    viewPDB_struct[iso].add_representation('cartoon', sele=':B', color=colorsB[9-iso])\n","\n","# Display the widget\n","viewPDB"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["The structures of the AC isoforms were created via homology modelling using the same template, so they are already aligned, as can be seen in the viewer above. The region where there are significant structural differences between the isoforms is in a flexible loop region that was not defined in the template structure. There are also variations in sequence length across AC isoforms in this region."]},{"metadata":{},"cell_type":"markdown","source":["## Calculating the electrostatic potentials of the proteins\n","\n","In the following cells, we calculate the electrostatic potentials of each AC isoform using the multipipsa software tool. For more information on this step, see the other molecular use case [Calculate the electrostatic potential of a protein from its atomic structure](https://collab.humanbrainproject.eu/#/collab/1655/nav/362934)."]},{"metadata":{"trusted":true,"collapsed":true},"cell_type":"code","source":["# Create a list of isoform structures\n","structures = [\"AC1\",\"AC2\", \"AC3\", \"AC4\", \"AC5\", \"AC6\", \"AC7\", \"AC8\", \"AC9\"]\n","# Define the location of the PIPSA software exectutables\n","pipsaDir = os.path.join(os.path.dirname(inspect.getfile(PipsaRun)), 'data', 'pipsa')"],"execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":["# Create an ApbsRun instance for the current calculation\n","epCalc = ApbsRun(\n","                    dataDir=useCaseDir,    # Pass the use case work directory as the directory for running the calculation\n","                    pipsaRoot=pipsaDir,    # Pass the location of the PIPSA executables defined above\n","                    temp='298.15',         # Define the temperature in Kelvin\n","                    ios='0.100',           # Define the solvent ionic strength in Molar concentration\n","                    pH='7.4',              # Define the solvent pH\n","                    structures=structures  # Pass the list of structures defined above\n","                ) \n","\n","epCalc.runPdb2Pqr()\n","epCalc.runApbs()"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["## Performing PIPSA analysis\n","\n","In this use case, we compare the electrostatic potentials of the AC isoforms within the whole skin surrounding the isoform structures. In the following cell we define a filename for our output results files, and initialise the calculation by creating an instance of the PipsaRun class in multipipsa."]},{"metadata":{"trusted":true,"collapsed":true},"cell_type":"code","source":["# Define an string for the output image files\n","imageFilename=\"wholeStructure\"\n","\n","pipsaCalc = PipsaRun(pipsaRoot=pipsaDir,\n","                     dataDir=useCaseDir,\n","                     pointsTemplate='AC5')"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["The following cell calculates all pairwise similarities for the electrostatic potentials of the AC isoforms. These are then used to cluster the isoforms."]},{"metadata":{"trusted":true},"cell_type":"code","source":["cluster = ClusterPipsa(structures=structures,\n","                       pipsaRoot=pipsaDir, \n","                       dataDir=useCaseDir, \n","                       distanceType=DistanceType.PIPSA,\n","                       graphicsFileRoot=imageFilename)\n","\n","pipsaCalc.runClusterPipsa(structures=structures,\n","                          points=[],\n","                          cluster=cluster)"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["## Visualizing the electrostatic similarity across isoforms \n","\n","The PIPSA analysis above creates as output an image file showing the pairwise distances between AC isoforms as a 2D heatmap. The results are also clustered using a single linkage hierachical method. The resulting dendograms are shown along the edges of the heatmap. The following cell displays this image"]},{"metadata":{"trusted":true},"cell_type":"code","source":["Image.open(os.path.join(useCaseDir, 'wholeStructure_heat3_1.png'))"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["The clustering obtained here can be compared to that in Figure 2A of [Tong et al (2016)](https://doi.org/10.1002/prot.25167) (although an unrooted dendrogram is shown there). The electrostatic similarity also reproduces some of the known regulation groupings of AC isoforms (see Table 1 in [Tong et al (2016)](https://doi.org/10.1002/prot.25167)). For example, clustering AC2, AC4 and AC7 together, and AC5 and AC5 together."]},{"metadata":{},"cell_type":"markdown","source":["## Saving your data to the collab storage area \n","In the final cell, your data will be moved to the storage area for your collab, from where you can download your files, and the local working directory will be cleaned."]},{"metadata":{"trusted":true,"collapsed":true},"cell_type":"code","source":["# Set up a timestamped directory name for saving results to the storage area\n","baseStorageDir = 'multipipsaWholePIPSA_'\n","timestamp = datetime.datetime.now().strftime('%Y-%m-%d-%H-%M-%S')\n","storageDir = os.path.join(collab_path, baseStorageDir + timestamp)\n","try:\n","    print('Creating storage directory: %s' % storageDir)\n","    storage_client.mkdir(storageDir)\n","except:\n","    print('There was an error creating the storage directory')\n","else:\n","    # Copy files to the storage area and remove the local files\n","    cleanDir = True\n","    for fName in os.listdir(useCaseDir):\n","        localFile = os.path.join(useCaseDir, fName)\n","        storageFile = os.path.join(storageDir, fName)\n","        fType = magic.Magic(mime=True).from_file(localFile)\n","        try:\n","            storage_client.upload_file(localFile, storageFile, fType)\n","        except:\n","            print('Error copying %s to storage' % fName)\n","            cleanDir = False\n","        else: \n","            os.remove(localFile)\n","            \n","    print('All files in the working directory have been moved to the storage area directory:')\n","    print(storageDir)\n","    os.chdir(homeDir)\n","    if cleanDir:\n","        os.rmdir(useCaseDir)"],"execution_count":null,"outputs":[]},{"metadata":{"trusted":true,"collapsed":true},"cell_type":"code","source":"","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"name":"python3","display_name":"Python 3","language":"python"},"language_info":{"name":"python","version":"3.6.7","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat":4,"nbformat_minor":2}